### DISCUSSION TITLE
Is there a function to create a folder if it doesn't exist?

### DISCUSSION THREAD

**Discussion by jonwest** [SilentVoid13/Templater#515] [Help]
To give some context, I'm also using the [Kanban plugin](https://github.com/mgmeyers/obsidian-kanban), and what I'd like to do is:

- Have a top-level "Projects" file that is displayed as a Kanban board.
- New notes created from that board would be created like `/Projects/<Project Title>.md`
- Then, the template for notes within that board would be:
```
---
kanban-plugin: basic
new-note-folder: /Projects/(PROJECT TITLE)/
---
```

Is it possible to have Templater create that folder?  Essentially this would allow me to generate a kanban board for each of my projects, but still keep notes and attachment from those projects separated from each other.

**Comment by welpdx** [comment]
Hey @jonwest 
Yes! There is a way to create a new folder!

You can use the `createFolder()` function in the Obsidian Api. ([It is referenced on line 1004](https://github.com/obsidianmd/obsidian-api/blob/bde556afa033e909ebfb9fcee8f5ef288276f78f/obsidian.d.ts#L3004))

You can check if path exists. If it doesn't, do:
```
await this.app.vault.createFolder("/foo/bar/")
```

![gif](https://i.imgur.com/ywJF5hx.gif)

**Comment by e-pet** [comment]
For future people like me with zero JS experience, this is a full snippet that checks whether a directory exists, creates the directory if it doesn't, and moves a file to it:
```
<%*
	const dir = "folder/folder/something/"
	if (!tp.file.exists(dir)) {
		await this.app.vault.createFolder(dir)
	}
	await tp.file.move(dir + tp.file.title)
%>
```

**Comment by welpdx** [comment]
Thanks @e-pet!
A similar snippet can be seen [here](https://github.com/SilentVoid13/Templater/discussions/333#discussioncomment-1302141) as well, from [this discussion](https://github.com/SilentVoid13/Templater/discussions/333).

However, please note that there is an issue with templater scripts not working properly when using the `await tp.file.move()` function. See [issue](https://github.com/SilentVoid13/Templater/issues/318#issuecomment-993007680) #318. If any one has a problem with their templates not working, refer to that [thread](https://github.com/SilentVoid13/Templater/issues/318#issuecomment-993007680) please.

**Comment by e-pet** [comment]
Ahh thanks @welpdx, I had not found that other thread! 

I actually just today got a "Template parsing error, aborting. Folder already exists." using the above snippet; would that likely be related to the thread you linked? Because it does sound like it's more related to the `this.app.vault.createFolder` than `tp.file.move`, right?

(While I'm already stealing your time, is there a way I can see which line in my template exactly threw the error? I know how I can open the console, but there it just says "main.ts:213" and if I click on that I end up in some main script with significantly less than 213 lines...?)

**Comment by welpdx** [comment]
Hi @e-pet,
Sorry for the late reply.

When an error occurs, I find that the console spits out a bunch of crap that doesnt help me at all. So, I usually debug my code with adding my own `console.log()` messages, `new Notice("Error occured",2000)` (see [Notice](https://github.com/obsidianmd/obsidian-api/blob/ddd50214f530d23ee21f440a9fa64f4507176871/obsidian.d.ts#L1991) in obs api), or some kind of try catch. 

Merry Christmas!

**Comment by GeniusBroccoli** [comment]
How to create folder structure in current folder?

**Reply by AB1908** [reply]
Where are you stuck and what have you tried?

**Reply by GeniusBroccoli** [reply]
how can I get folder of current note?

**Reply by GeniusBroccoli** [reply]
okay, thx for help, I made it
```
const folders = ["attachments", "attachments/docs/", "attachments/images/"]
const actFolder = this.app.workspace.getActiveFile().parent.path
folders.forEach(folder => {
	if (!!tp.file.exists(`${actFolder}/${folder}`)) {
		this.app.vault.createFolder(`${actFolder}/${folder}`)
	}
})
```

This creates attachment folder in folder with active md file and adds docs and images subfolders. In future I want edit this to create subfolders in current folder depending on the types of files and move that files to created folders. I don't always attach files to notes, and this will allow me to categorize files into subfolders by type.
